File: sdk/php/CertdClient.php
<?php

class CertdClient
{
    private string $keyId;
    private string $keySecret;
    private string $baseUrl;
    private bool $encrypt;
    private string $signType;

    public function __construct(string $keyId, string $keySecret, array $options = [])
    {
        if ($keyId === '') {
            throw new InvalidArgumentException('keyId is required');
        }
        if ($keySecret === '') {
            throw new InvalidArgumentException('keySecret is required');
        }
        $this->keyId = $keyId;
        $this->keySecret = $keySecret;
        $this->baseUrl = rtrim($options['baseUrl'] ?? 'http://127.0.0.1:7001', '/');
        $this->encrypt = $options['encrypt'] ?? false;
        $this->signType = $options['signType'] ?? 'md5';
    }

    public function getSign(string $content): string
    {
        if ($this->signType !== 'md5') {
            throw new InvalidArgumentException("Unsupported signType: {$this->signType}");
        }
        return md5($content . $this->keySecret);
    }

    public function getToken(?bool $encrypt = null): string
    {
        $content = json_encode([
            'keyId' => $this->keyId,
            't' => time(),
            'encrypt' => $encrypt ?? $this->encrypt,
            'signType' => $this->signType,
        ], JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        $sign = $this->getSign($content);
        return base64_encode($content) . '.' . base64_encode($sign);
    }

    public function request(string $path, array $body = [], ?bool $encrypt = null): string
    {
        if (!function_exists('curl_init')) {
            throw new RuntimeException('PHP curl extension is required');
        }

        $payload = json_encode($body, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        $headers = [
            'Content-Type: application/json',
            'x-certd-token: ' . $this->getToken($encrypt),
        ];

        $ch = curl_init($this->baseUrl . $path);
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => $payload,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 60,
        ]);

        $response = curl_exec($ch);
        if ($response === false) {
            throw new RuntimeException(curl_error($ch));
        }
        $statusCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($statusCode < 200 || $statusCode >= 300) {
            throw new RuntimeException("HTTP {$statusCode}: {$response}");
        }
        return $response;
    }

    public function getCert(array $params): string
    {
        return $this->request('/api/v1/cert/get', $params);
    }
}

File: sdk/python/certd_client.py
import base64
import hashlib
import json
import time
import urllib.error
import urllib.request
from typing import Any, Dict, Optional


class CertdClient:
    def __init__(
        self,
        key_id: str,
        key_secret: str,
        base_url: str = "http://127.0.0.1:7001",
        encrypt: bool = False,
        sign_type: str = "md5",
    ) -> None:
        if not key_id:
            raise ValueError("key_id is required")
        if not key_secret:
            raise ValueError("key_secret is required")
        self.key_id = key_id
        self.key_secret = key_secret
        self.base_url = base_url.rstrip("/")
        self.encrypt = encrypt
        self.sign_type = sign_type

    def get_sign(self, content: str) -> str:
        if self.sign_type != "md5":
            raise ValueError(f"Unsupported sign_type: {self.sign_type}")
        return hashlib.md5((content + self.key_secret).encode("utf-8")).hexdigest()

    def getSign(self, content: str) -> str:
        return self.get_sign(content)

    def get_token(self, encrypt: Optional[bool] = None) -> str:
        content = json.dumps(
            {
                "keyId": self.key_id,
                "t": int(time.time()),
                "encrypt": self.encrypt if encrypt is None else encrypt,
                "signType": self.sign_type,
            },
            separators=(",", ":"),
            ensure_ascii=False,
        )
        sign = self.get_sign(content)
        return (
            base64.b64encode(content.encode("utf-8")).decode("ascii")
            + "."
            + base64.b64encode(sign.encode("utf-8")).decode("ascii")
        )

    def getToken(self, encrypt: Optional[bool] = None) -> str:
        return self.get_token(encrypt)

    def request(self, path: str, body: Optional[Dict[str, Any]] = None, encrypt: Optional[bool] = None) -> str:
        data = json.dumps(body or {}, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
        request = urllib.request.Request(
            f"{self.base_url}{path}",
            data=data,
            method="POST",
            headers={
                "content-type": "application/json",
                "x-certd-token": self.get_token(encrypt),
            },
        )

        try:
            with urllib.request.urlopen(request, timeout=60) as response:
                return response.read().decode("utf-8")
        except urllib.error.HTTPError as error:
            message = error.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"HTTP {error.code}: {message}") from error

    def get_cert(self, params: Dict[str, Any]) -> str:
        return self.request("/api/v1/cert/get", params)

    def getCert(self, params: Dict[str, Any]) -> str:
        return self.get_cert(params)
